_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+}
+```
+Fundementally, this idea is very straightforward and not too difficult to
+implement. Right? _Right?_
+
+## Challenge 1: Lifetime Coercion
+
+Unfortunately, not all code lives on this happy path. Check this out:
+```rust
+// `Foo` and `Bar` same as above
+impl<'a> Foo<'a> {
+    pub fn get_bar<'b>(&self) -> &'b Bar
+    where
+        'a: 'b,
+    {
+        &self.bar
+    }
+}
+```
+To be honest, I've never had to use lifetime bounds outside of borrowing
+`Deserialize` impls, and Diplomat doesn't support generic types, so I'm not sure
+why you'd need bounded lifetimes. But they exist, and I didn't want to leave it
+as someone else's problem.
+
+The problem here is that the output looks like this:
+```rust
+&'b Bar
+```
+While the input looks like this:
+```rust
+self: &Foo<'a>
+```
+**These are not the same lifetime**. If you just look in the input for `'b`,
+you'll find that it doesn't borrow from the input, and this can and will
+eventually lead to dangling pointers. But this works in Rust because the
+`'a: 'b` bound allows`'a` to coerce into `'b`. Therefore, we have to be
+cognizant of these bounds as well. But it goes even further:
+```rust
+impl<'a> Foo<'a> {
+    pub fn get_bar<'b, 'c>(&self) -> &'c Bar
+    where
+        'a: 'b,
+        'b: 'c,
+    {
+        &self.bar
+    }
+}
+```
+Continuing the theme from above, I have no idea why you would ever write this,
+especially in a set of FFI bindings. But it's legal Rust code and I wanted to
+be consistent with the language.
+
+Here, the output has lifetime `'c`, and the input has lifetime `'a`, but it
+doesn't explicitly say `'a: 'c`. Instead, it's _inferred_ via transitivity.
+Do you see where this is going?
+
+Here's an exaggurated example of a `where` clause:
+```rust
+'a: 'b,
+'b: 'c,
+'c: 'e,
+'d: 'b,
+'e: 'd + 'f,
+```
+This looks lot like an adjacency list. That's right, this is a directed graph
+problem.
+
+If you think of lifetimes as vertices and bounds as edges pointing from longer
+to shorter lifetimes, then determining possible coercions becomes a
+simple reachability problem. The diagram below shows the exaggurated example
+as a directed graph:
+<figure>
+    <p align="center">
+        <img src="{{site.baseurl}}/images/google-internship/directed-graph.png" alt="A black-on-white image of a directed graph with 6 vertices labeled A through F. A points to B, B points to C, C points to E, E points to D and F, and D points back to B. It clearly shows that B, C, E, and D form a cycle." width="400"/>
+        <figcaption>Graph representation of lifetimes and bounds</figcaption>
+    </p>
+</figure>
+
+With this model, we can easily generalize the definition of what it means for
+some lifetime A to be able to coerce into some lifetime B. That is, lifetime A
+can coerce into lifetime B if and only if there exists a path from A to B on
+the directed graph of lifetimes and bounds. In practice, this also mean that
+all lifetimes in a cycle are all the same lifetime.
```

Comment:
q: can you even do cycles with lifetimes in Rust?

Reply:
yup
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=248c983a65fa835ea9e87be763aeec88